core/emulate.c

```diff
@@ -292,6 +310,58 @@ static const struct em_opcode_t opcode_table_0F3A[256] = {
     X16(N), X16(N), X16(N), X16(N),
 };
 
+static uint64_t gpr_read_shifted(struct em_context_t *ctxt,
+                                 unsigned index, size_t size, size_t shift)
+{
+    uint64_t value = 0;
+    uint8_t *value_ptr;
+
+    if (!(ctxt->gpr_cache_r & (1 << index))) {
+        ctxt->gpr_cache[index] = ctxt->ops->read_gpr(ctxt->vcpu, index);
+        ctxt->gpr_cache_r |= (1 << index);
+    }
+    value_ptr = (uint8_t*)&ctxt->gpr_cache[index];
+    memcpy(&value, &value_ptr[shift], size);
+    return value;
+}
+
+static void gpr_write_shifted(struct em_context_t *ctxt,
+                              unsigned index, size_t size, size_t value, uint64_t shift)
+{
+    uint8_t *value_ptr;
+
```

Comment:
nit: Add `assert()` statements to detect potential buffer overflows:

* `index < 16`
* `size + shift <= 8`